hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    evtHandler.handler = cb;
+    evtHandler.context = context;
+    genericEventHandlers_.append(evtHandler);
+}
+
+os_thread_return_t BleObject::BleEventDispatcher::bleEventDispatch(void* param) {
+    BleEventDispatcher* dispatcher = static_cast<BleEventDispatcher*>(param);
+    while (1) {
+        EventMessage msg;
+        if (!os_queue_take(dispatcher->evtQueue_, &msg, CONCURRENT_WAIT_FOREVER, nullptr)) {
+            if (msg.handler) {
+                // Invoke the handler registered by specific application.
+                if (msg.evt.type == BLE_EVT_SCAN_RESULT) {
+                    msg.handler(&msg.evt.params.scan_result, msg.context);
+                } else if (msg.evt.type == BLE_EVT_SVC_DISCOVERED) {
+                    msg.handler(&msg.evt.params.svc_disc, msg.context);
+                } else if (msg.evt.type == BLE_EVT_CHAR_DISCOVERED) {
+                    msg.handler(&msg.evt.params.char_disc, msg.context);
+                }
+            } else {
+                // Just dispatch the event to the application those have subscribed the generic events.
+                for (int i = 0; i < dispatcher->genericEventHandlers_.size(); i++) {
+                    BleGenericEventHandler& evtHandler = dispatcher->genericEventHandlers_[i];
+                    if (evtHandler.handler) {
+                        evtHandler.handler(&msg.evt, evtHandler.context);
+                    }
+                }
+            }
+            // Free the memory if needed.
+            if (msg.hook) {
+                msg.hook(&msg.evt, msg.hookContext);
+            }
         } else {
-            // Pairing is not supported
-            ret = sd_ble_gap_sec_params_reply(conn, BLE_GAP_SEC_STATUS_PAIRING_NOT_SUPP, nullptr, nullptr);
-        }
-        if (ret != NRF_SUCCESS) {
-            LOG(ERROR, "sd_ble_gap_sec_params_reply() failed: %u", (unsigned)ret);
+            LOG(ERROR, "BLE event thread exited.");
+            break;
         }
-        break;
     }
-    case BLE_GAP_EVT_AUTH_STATUS: {
-        const auto status = event->evt.gap_evt.params.auth_status.auth_status;
-        if (status == BLE_GAP_SEC_STATUS_SUCCESS) {
-            LOG_DEBUG(TRACE, "Authentication succeeded");
-        } else {
-            LOG(WARN, "Authentication failed, status: %d", (int)status);
-        }
-        break;
+    os_thread_exit(dispatcher->evtThread_);
+}
+
+typedef struct {
+    BleObject::BleGap* instance;
+} BleGapImpl;
+static BleGapImpl bleGapImpl;
+
+BleObject::BleGap::BleGap(BleObject* obj) : bleObj_(obj) {
+}
+
+int BleObject::BleGap::init(void) {
+    // Set the default device name
+    char devName[32] = {};
+    get_device_name(devName, sizeof(devName));
+    setDeviceName((const uint8_t*)devName, strlen(devName));
+    bleGapImpl.instance = this;
+    NRF_SDH_BLE_OBSERVER(bleGap, 1, processBleGapEvents, &bleGapImpl);
+    return SYSTEM_ERROR_NONE;
+}
+
+int BleObject::BleGap::setDeviceName(const uint8_t* deviceName, uint16_t len) {
```

Comment:
Any reason this cannot simply be `const char* deviceName` ? It is a string after all.